File: AlgorithmsPackPythonSolutions/416. Partition Equal Subset Sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        MAX = 200 + 1
        MAX_T = 100 * 200 + 1
        memory = [[-1] * MAX_T for _ in range(MAX)]
        numms = nums

        def partition(index, target):
            if target < 0:
                return False

            if target == 0:
                return True

            if index >= len(numms):
                return False

            if memory[index][target] != -1:
                return memory[index][target]

            if partition(index + 1, target):
                return True

            memory[index][target] = partition(index + 1, target - numms[index])
            return memory[index][target]

        def canPart(nums, target):
            nonlocal numms
            numms = nums
            return partition(0, target)

        total_sum = sum(nums)
        if total_sum % 2 == 1:
            return False

        return canPart(nums, total_sum // 2)

File: AlgorithmsPackJavaSolutions/416. Partition Equal Subset Sum.java
class Solution {
    private int[][] memory;
    private int[] numms;

    public boolean canPartition(int[] nums) {
        int MAX = 200 + 1;
        int MAX_T = 100 * 200 + 1;
        memory = new int[MAX][MAX_T];
        numms = nums;

        for (int[] row : memory) {
            Arrays.fill(row, -1);
        }

        int totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }

        if (totalSum % 2 == 1) {
            return false;
        }

        return canPart(nums, totalSum / 2);
    }

    private boolean partition(int index, int target) {
        if (target < 0) {
            return false;
        }

        if (target == 0) {
            return true;
        }

        if (index >= numms.length) {
            return false;
        }

        if (memory[index][target] != -1) {
            return memory[index][target] == 1;
        }

        if (partition(index + 1, target)) {
            memory[index][target] = 1;
            return true;
        }

        memory[index][target] = partition(index + 1, target - numms[index]) ? 1 : 0;
        return memory[index][target] == 1;
    }

    private boolean canPart(int[] nums, int target) {
        numms = nums;
        return partition(0, target);
    }
}
